lib/sdotypes.c

```diff
@@ -3569,6 +3658,609 @@ bool sdo_encrypted_packet_windup(sdow_t *sdow, int type, sdo_iv_t *iv)
+	}
+	ret = true;
+end:
+	sdo_byte_array_free(empty_uph);
+	return ret;
+}
+
+/**
+ * Read the given COSE into the fdo_cose_t parameter.
+ * The fdo_cose_t parameter should have memory pre-allocated.
+ * However, the internal elements must be un-allocated.
+ * The memory allocation for the same would be done in the method.
+ * [
+ * protected header,
+ * unprotected header,
+ * payload,				// bstr
+ * signature			// bstr
+ * ]
+ * Return true, if read was a success. False otherwise.
+ */
+bool fdo_cose_read(sdor_t *sdor, fdo_cose_t *cose) {
+
+	size_t num_cose_items = 4;
+	if (!sdor_array_length(sdor, &num_cose_items) || num_cose_items != 4) {
+		LOG(LOG_ERROR, "COSE: Failed to read/Invalid array length\n");
+		return false;		
+	}
+
+	if (!sdor_start_array(sdor)) {
+		LOG(LOG_ERROR, "COSE: Failed to read start array\n");
+		return false;
+	}
+
+	cose->cose_ph = sdo_alloc(sizeof(fdo_cose_protected_header_t));
+	if (!cose->cose_ph) {
+		LOG(LOG_ERROR, "COSE: Failed to alloc Protected Header\n");
+		goto end;
+	}
+	if (!fdo_cose_read_protected_header(sdor, cose->cose_ph)) {
+		LOG(LOG_ERROR, "COSE: Failed to read protected header\n");
+		goto end;
+	}
+
+	if (!fdo_cose_read_unprotected_header(sdor)) {
+		LOG(LOG_ERROR, "COSE: Failed to read unprotected header\n");
+		goto end;
+	}
+
+	size_t var_length = 0;
+	if (!sdor_string_length(sdor, &var_length) ||
+		var_length == 0) {
+		LOG(LOG_ERROR, "COSE: Failed to read payload length\n");
+		goto end;	
+	}
+	cose->cose_payload = sdo_byte_array_alloc(var_length);
+	if (!cose->cose_payload) {
+		LOG(LOG_ERROR, "COSE: Failed to alloc EATPayload\n");
+		goto end;
+	}
+	if (!sdor_byte_string(sdor, cose->cose_payload->bytes, cose->cose_payload->byte_sz)) {
+		LOG(LOG_ERROR, "COSE: Failed to read payload\n");
+		goto end;
+	}
+
+	var_length = 0;
+	if (!sdor_string_length(sdor, &var_length) ||
+		var_length == 0) {
+		LOG(LOG_ERROR, "COSE: Failed to read signature length\n");
+		goto end;	
+	}
+	cose->cose_signature = sdo_byte_array_alloc(var_length);
+	if (!cose->cose_signature) {
+		LOG(LOG_ERROR, "COSE: Failed to alloc Signature\n");
+		goto end;
+	}
+	if (!sdor_byte_string(sdor, cose->cose_signature->bytes, cose->cose_signature->byte_sz)) {
+		LOG(LOG_ERROR, "COSE: Failed to read signature\n");
+		goto end;
+	}
+
+	if (!sdor_end_array(sdor)) {
+		LOG(LOG_ERROR, "COSE: Failed to read end array\n");
+		goto end;
+	}
+	return true;
+
+end:
+	fdo_cose_free(cose);
+	return false;
+}
+
+void fdo_rvto2addr_entry_free(fdo_rvto2addr_entry_t *rvto2addr_entry) {
+	if (rvto2addr_entry->rvip)
+		sdo_byte_array_free(rvto2addr_entry->rvip);
+	if (rvto2addr_entry->rvdns)
+		sdo_string_free(rvto2addr_entry->rvdns);
+	sdo_free(rvto2addr_entry);	
+}
+
+void fdo_rvto2addr_free(fdo_rvto2addr_t *rvto2addr) {
+	if (rvto2addr) {
+		while (rvto2addr->rv_to2addr_entry) {
+			fdo_rvto2addr_entry_free(rvto2addr->rv_to2addr_entry);
+			rvto2addr->rv_to2addr_entry = (fdo_rvto2addr_entry_t *) rvto2addr->rv_to2addr_entry->next;
```

Comment:
In line 4115, are we not trying access an object that is already freed?

Suggested alternative:
```
		while (rvto2addr->rv_to2addr_entry) {
			fdo_rvto2addr_entry_t *rv_to2addr_entry = rvto2addr->rv_to2addr_entry;
			rvto2addr->rv_to2addr_entry = rvto2addr->rv_to2addr_entry->next;
			fdo_rvto2addr_entry_free(rv_to2addr_entry);			
		}
```

Reply:
Done.